text/0000-address.md

```diff
@@ -0,0 +1,231 @@
+that he has sent a token to the public key.
+
+To create a new address, we only need the following information:
+
+* `PublicKey`: this will be necessary to be able to utilise the token;
+* `Discriminant`: this will prevent misuse of addresses between testing
+  environment and production environment;
+
+This is the most simple address. It is called a **Single Address**. This is
+because it is not associated with any staking key.
+
+It is possible to create a **Grouped Address**, it groups addresses to a
+staking key. To create a **Grouped Address** you need the `PublicKey`,
+the `Discriminant` (just like for a **Single Address**) and another
+`PublicKey` that is associating this address to another hierarchy.
+
+The human readable encoding of the `Address` is [bech32]. It includes
+a checksum for correctness (preventing mistyping the address). There
+is also a visual discriminant for human readable discrimination between
+cardano (production) addresses and testing version of the blockchain:
+
+* `ta` is the visual discriminant for **Testing** address;
+* `ca` is the visual discriminant for **Cardano** address (production address);
+
+So to create a **Single Address** we need a public key.
+
+```
+PublicKey = 0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20
+```
+
+A production **Single Address** would look like:
+`ca1qvqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jqxuzx4s`. Note the `ca` at
+the beginning of the `Address`. This is our visual discriminant.
+
+To create a **Grouped Address** we need an additional grouping key:
+
+```rust
+GroupKey = 292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f404142434445464748
+```
+
+A production **Grouped Address** would look like:
+`ca1qsqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jq2f29vkz6t30xqcnyve5x5mrwwpe8ganc0f78aqyzsjrg3z5v36gguhxny`.
+Note the `ca` prefix again.
+
+A test **Grouped Address** would look like:
+`ta1ss5j52ev95hz7vp3xgengdfkxuurjw3m8s7nu06qg9pyx3z9ger5sqgzqvzq2ps8pqys5zcvp58q7yq3zgf3g9gkzuvpjxsmrsw3u8eqx5x7xh`.
+Note the `ta` prefix here.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Binary serialization
+
+It uses a simple serialization format which is made to be concise:
+
+* First byte contains the discrimination information (1 bit) and the kind of address (7 bits)
+* Remaining bytes contains a kind specific encoding describe after.
+
+2 kind of address are currently supported:
+
+* Single: Just a (spending) public key using the ED25519 algorithm
+* Group: Same as single, but with a added (staking/group) public key
+  using the ED25519 algorithm.
+
+```
+Single key:
+    DISCRIMINATION_BIT || SINGLE_KIND_TYPE (7 bits) || SPENDING_KEY
+
+Group key:
+    DISCRIMINATION_BIT || GROUP_KIND_TYPE (7 bits)|| SPENDING_KEY || STAKING_KEY
+```
+
+Bit value for the `DISCRIMINATION_BIT`:
+
+| value | purpose |
+|---:|:--- |
+| 0 | Production Address |
+| 1 | Test Address |
```

Comment:
Is two networks enough? E.g. for Cardano alone we already have `mainnet`, `testing` and `staging`, and you might want to prevent address reuse between alternative Cardano-based chains as well...

Reply:
Ah, I see you addressed this elsewhere.